microbench/src/main/java/io/netty/microbenchmark/common/IsValidIpV4Benchmark.java

```diff
@@ -0,0 +1,125 @@
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Threads;
+import org.openjdk.jmh.annotations.Warmup;
+
+import java.util.concurrent.TimeUnit;
+
+@Threads(1)
+@Warmup(iterations = 3)
+@Measurement(iterations = 3)
+@OutputTimeUnit(TimeUnit.MICROSECONDS)
+public class IsValidIpV4Benchmark extends AbstractMicrobenchmark {
+
+    @Param({ "127.0.0.1", "255.255.255.255", "1.1.1.1", "127.0.0.256", "127.0.0.1.1", "127.0.0", "[2001::1]" })
+    private String ip;
+
+    // Old
+
+    public static boolean isValidIpV4AddressOld(String value) {
+
+        int periods = 0;
+        int i;
+        int length = value.length();
+
+        if (length > 15) {
+            return false;
+        }
+        char c;
+        StringBuilder word = new StringBuilder();
+        for (i = 0; i < length; i ++) {
+            c = value.charAt(i);
+            if (c == '.') {
+                periods ++;
+                if (periods > 3) {
+                    return false;
+                }
+                if (word.length() == 0) {
+                    return false;
+                }
+                if (Integer.parseInt(word.toString()) > 255) {
+                    return false;
+                }
+                word.delete(0, word.length());
+            } else if (!Character.isDigit(c)) {
+                return false;
+            } else {
+                if (word.length() > 2) {
+                    return false;
+                }
+                word.append(c);
+            }
+        }
+
+        if (word.length() == 0 || Integer.parseInt(word.toString()) > 255) {
+            return false;
+        }
+
+        return periods == 3;
+    }
+
+    // New
+
+    private static boolean isValidIpV4Word(CharSequence word, int start, int endExclusive) {
+        int len = endExclusive - start;
+        char c0, c1, c2;
+        if (len < 1 || len > 3 || (c0 = word.charAt(start)) < '0') {
+            return false;
+        }
+        if (len == 3) {
+            return (c1 = word.charAt(start + 1)) >= '0' &&
+                   (c2 = word.charAt(start + 2)) >= '0' &&
+                   (c0 <= '1' && c1 <= '9' && c2 <= '9' ||
+                    c0 == '2' && c1 <= '5' && (c2 <= '5' || c1 < '5' && c2 <= '9'));
+        }
+        return c0 <= '9' && (len == 1 || isValidNumericChar(word.charAt(start + 1)));
+    }
+
+    private static boolean isValidNumericChar(char c) {
+        return c >= '0' && c <= '9';
+    }
+
+    @SuppressWarnings("DuplicateBooleanBranch")
+    public static boolean isValidIpV4AddressNew(String ip) {
+        int len = ip.length();
+        int i;
+        int j;
+        return len <= 15 && len >= 7 &&
+               (i = ip.indexOf('.', 0)) > 0 && isValidIpV4Word(ip, 0, i) &&
+               (i = ip.indexOf('.', j = i + 2)) > 0 && isValidIpV4Word(ip, j - 1, i) &&
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Extract the assignment out of this expression. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AAssignmentInSubExpressionCheck)